### PR TITLE
A version of automodapi for PlasmaPy

### PR DESCRIPTION
**Pull request by rocco8773**
This PR creates the sphinx extension package `plasmapy_sphinx`.  This package defines two reST directives `automodapi` and `automodsumm`, which are an adaption of the likely named directives from `sphinx_automdoapi`.  There are a few significant distinctions of our directives over the `sphinx_automodapi` directives that made me add the code here and not push it upstream to `sphinx_automodapi`.

1. I introduced the concept of groups for the directives.  So, when an module is inspected for its objects, the found objects are categorized into group types (modules, functions, classes, exceptions, etc.).  As a result, the `sphinx_automodapi` options like `:functions-only:` are replaced by `:groups: functions`.
2. This move to the groups options allows for multiple groups to be displayed at once for `automodsumm` by doing

    ``` rst
    .. automodsumm:: foo.bar
       :groups: functions, classes

    .. automodapi:: foo.bar
       :groups: functions, classes
    ```

    It additionally allows for custom groups to be defined.
3. Configuration values (that are defined in ``conf.py``) are introduced that allows you define custom groups.  To define a custom group the `automodapi_custom_groups` configuration value needs to be defined like

    ```python
    automodapi_custom_groups = {
        "mygroup" : {
            "title": "My Custum Group",
            "dunder": "__mygroup__",
        },
    }
    ```

    In this example `mygroup` would be the name of the custom group, `"title"` would be the title rendered above the `autosummary` table when `automodapi` is used, and `"dunder"` is the name of the dunder variable used to list the custom group objects.  This `__mygroup__` would be placed in the module being documented with the custom group and is defined just like `__all__`.
4. The `automodapi` directive is an official `autodoc` directive and not a function connected to the [`source-read`](https://www.sphinx-doc.org/en/master/extdev/appapi.html?highlight=source-read#event-source-read) event.  This allows `automodapi` to benefit from all the default functionality associated with the `autodoc` directives.

### CODE REVIEW

File: docs/automodapi/__init__.py
Comment (StanczakDominik) [inline docs/automodapi/__init__.py:1]: Should note the license of automodapi here! Also, dev0 is probably autogenerated and it'd be better to point to a particular commit.

**Review by StanczakDominik — COMMENTED** [review]
The mobile app is not letting me do a single line comment for some reason.

**Comment by rocco8773** [comment]
@StanczakDominik I'm not exactly sure why the tests are failing.  I believe the runner currently thinks it has to run tests on any `.py` file in `docs/`  and `sphinx` isn't included in the dependencies for those test environments, so it fails.  I tried changing a few things in the `setup.cfg`, but with no success.  Do you mind looking into this?  I'm going to be out of town through the weekend so there's no rush.  Thanks.

**Comment by StanczakDominik** [comment]
Sure, I'll get them passing :) don't worry about it.

File: docs/automodapi/__init__.py
Comment (rocco8773) [reply docs/automodapi/__init__.py:1]: There is still a lot of clean up to be had here and there, including properly linking to `sphinx_automodapi`.  I just want to get the bulk working and then I'll go back to dot the i's and cross the t's.

**Comment by StanczakDominik** [comment]
That should do the trick for tests. I think testing `docs` was a remnant of sphinx-gallery. No tests from docs were being collected by pytest, anyway, so I just removed that. From now on, all .py files within docs or its subdirectories will be ignored.

**Comment by StanczakDominik** [comment]
Darn, that worked. I hate it when that happens.

---

Well, it imports, at least, but throws warnings on stubfiles for formulary functions. That's no CI bug, so I'll leave that to the expert here! :wink:

**Comment by rocco8773** [comment]
@StanczakDominik The documentation CI is failing because the `tox` `testenv` is changing directories tot he root directory (where the `ini` is at) and this is causing my (our) `automodsumm` to generated stub files in the root directory instead of the `docs/` directory.  There are several ways of fixing this, but I'm going to try to do this through modifying `automodsumm`.

**Comment by StanczakDominik** [comment]
Or I can try looking at how to build the docs without changing directories? Won't that work as well? More fragile, though, probably.

**Comment by rocco8773** [comment]
It would and it would only require changing one line in the `tox.ini`. You just set `currentdir = {toxinidir}/docs`. However, `autosummary` and the original`automodsumm` are robust enough to deal with this, so ours should be as well. I already know where I need to make the change.

**Comment by StanczakDominik** [comment]
Hey, the flake8 failure stuff is new after #1062 got merged. Don't worry about it for now; I'll disable the requirement in branch protection.

**Comment by StanczakDominik** [comment]
Also,

image

do iiiiiiiiiiiiiiiiit

File: docs/conf.py
Comment (StanczakDominik) [inline docs/conf.py:29]: Like it or not, this is probably the simplest way of making a package living in `docs` be importable :(

**Comment by StanczakDominik** [comment]
I'm not seeing RTD in the checks tab but it seems to be building:

https://plasmapy--1105.org.readthedocs.build/en/1105/formulary/parameters.html

File: docs/conf.py
Comment (rocco8773) [reply docs/conf.py:29]: Yeah, this is weird.  I'm curious if it's related to the directory change done by `tox`.  Anyways, we will eventually break `automodapi` out into it's own package, so I'm not going to fret about it for now.

**Comment by rocco8773** [comment]
> Also,
> 
> image
> 
> do iiiiiiiiiiiiiiiiit

I know, I know.  I still have several things to do so I'm not worrying about linters at the moment.  I'll take care of that when I'm crossing the T's and dotting the i's.

**Comment by rocco8773** [comment]
This is mostly there, so I think it's ready for feedback.  The only major thing missing is adding docstrings to the new functionality.  The additional functionality I'd like to add is: (1) a `:group-order:` option for the `automodapi` directive so the display order of groups (modules, classes, functions, etc.) can be changed on the fly and (2) have the ability to add brief a description under the group heading and above the `autosummary` table that is defined by the `conf.py` config variable `automod_custom_groups`.  Both of these could come in a follow up PR, but I don't think they'll be hard to implement.

**Comment by rocco8773** [comment]
Sorting of the objects in the tables also appears to be off, so I'll need to look into that too.

**Comment by StanczakDominik** [comment]
Okay, since this is in the ready-to-review stage, I really want to try this out :D

pre-commit.ci autofix

**Review by StanczakDominik — APPROVED** [review]
Looks good in terms of results! I skimmed over the `plasmapy_sphinx` subpackage; I see the results are lovely and no crypto mining is going on in there, but that's the extent of what I can tell 😜 But I'd be happy to merge it with intent to read through the code itself when I finally start picking up sphinx.

**Comment by StanczakDominik** [comment]
pre-commit.ci autofix

File: docs/ad/analysis/swept_langmuir.rst
Comment (rocco8773) [reply docs/ad/analysis/swept_langmuir.rst:23]: `automodapi` is a "sub-class" of the `sphinx.ext.autodoc` directive `automodule` and, thus, the `automodule` options still work for `automodapi`.  The `sphinx.ext.autodoc` functionality is what sphinx uses to index the documented pacakage.  Basically this is what populates the `object.inv`, which is used by `sphinx.ext.intersphinx` to index and cross-link the documented package.  Since `.. automodule:: plasmapy.analysis.swept_langmuir` is used a few lines above here, without `:noindex:` sphinx would be trying to index the same package at two difference locations in the documentation.

**Comment by StanczakDominik** [comment]
As per telecon - still waiting for documentation for the merge. And for a thorough code review,

/remind me in 6 months

**Comment by rocco8773** [comment]
I'm gonna have a few follow up PRs to this, but, for now, I'm done.  So, take out your red-est pen and have at it!

**Comment by rocco8773** [comment]
Note: If you go to the modules section of the built RTD, then you'll find the documentation pages for `plasmapy_sphinx`.

File: docs/plasmapy_sphinx/extras.py
Comment (namurphy) [inline docs/plasmapy_sphinx/extras.py:2]:
```suggestion
This module contains extra (optional) Sphinx setup parameters that are not needed
```

File: docs/plasmapy_sphinx/extras.py
Comment (namurphy) [inline docs/plasmapy_sphinx/extras.py:3]:
```suggestion
to run the `plasmapy_sphinx` extension but are useful for documentation.  See
```
Not sure if this suggestion is along with what you intended.  Otherwise I was wondering if there was another word that would follow "documenting".

File: docs/plasmapy_sphinx/extras.py
Comment (namurphy) [inline docs/plasmapy_sphinx/extras.py:18]:
```suggestion
    Used to set up the ``event`` directive and role for documenting Sphinx events.
```

File: docs/plasmapy_sphinx/extras.py
Comment (namurphy) [inline docs/plasmapy_sphinx/extras.py:49]: If this were Python 3.8+ only, then this would have been a good place for a walrus `:=` operator.  Alas!  

[optional] I'm wondering if there would be a more descriptive variable name than `m`.  Not high priority, though.

File: docs/plasmapy_sphinx/extras.py
Comment (namurphy) [inline docs/plasmapy_sphinx/extras.py:62]:
```suggestion
    to set up convenient, but optional, functionality.  This adds:
```

File: docs/plasmapy_sphinx/utils.py
Comment (namurphy) [inline docs/plasmapy_sphinx/utils.py:66]:
```suggestion
        Name of the module (e.g. ``"plasmapy_sphinx.utils'``) to be inspect.
```

File: docs/plasmapy_sphinx/extras.py
Comment (rocco8773) [reply docs/plasmapy_sphinx/extras.py:49]: This was copied and pasted straight from sphinx's `conf.py`.  That's why it does not resemble my coding style.

File: docs/plasmapy_sphinx/__init__.py
Comment (namurphy) [inline docs/plasmapy_sphinx/__init__.py:16]:
```suggestion
is one of the extension mechanisms of reST and, thus, Sphinx.
```

File: docs/plasmapy_sphinx/__init__.py
Comment (namurphy) [inline docs/plasmapy_sphinx/__init__.py:24]:
```suggestion
Configuration values are variables that can be defined in the `conf.py` file
```

File: docs/plasmapy_sphinx/automodapi.py
Comment (namurphy) [inline docs/plasmapy_sphinx/automodapi.py:47]:
```suggestion
        | **exceptions** | Classes that inherit from `BaseException` (excluding      |
```

File: docs/plasmapy_sphinx/automodapi.py
Comment (namurphy) [inline docs/plasmapy_sphinx/automodapi.py:78]:
```suggestion
        The behavior of ``:no-main-docstring:`` is described below
        in the :rst:dir:`automodapi:no-main-docstring` section.  If you
```

File: docs/plasmapy_sphinx/automodapi.py
Comment (namurphy) [inline docs/plasmapy_sphinx/automodapi.py:122]:
```suggestion
        the object's short name.  Continuing with the example from above, let's
        skip `~plasmapy_sphinx.automodapi.ModAPIDocumenter` and
```

File: docs/plasmapy_sphinx/automodapi.py
Comment (namurphy) [inline docs/plasmapy_sphinx/automodapi.py:184]:
```suggestion
        used for the heading created by :rst:dir:`automodapi`.  The code

        .. code-block:: rst

            .. automodapi:: foo.bar
               :include-heading:
               :heading-chars: ^~

        would generate reStructuredText equivalent to
```

File: docs/plasmapy_sphinx/automodapi.py
Comment (namurphy) [inline docs/plasmapy_sphinx/automodapi.py:193]: 🙀  🙀  🙀  🙀  🙀  🙀  😹

File: docs/plasmapy_sphinx/automodapi.py
Comment (namurphy) [inline docs/plasmapy_sphinx/automodapi.py:268]:
```suggestion
A configuration value is a variable that can be defined in ``conf.py`` to configure
```

File: docs/plasmapy_sphinx/automodapi.py
Comment (namurphy) [inline docs/plasmapy_sphinx/automodapi.py:266]: There is a lot of useful functionality in this section!

**Review by namurphy — COMMENTED** [review]
A partial review focusing on docstrings in `__init__.py` and `automodapi.py`.  Looks like there is some really useful functionality in there.  The suggestions are pretty minor.  In a few places I wasn't sure if my suggestion referred to what you intended so feel free to adjust as needed.  And thank you again for doing this!

**Comment by rebeccaringuette** [comment]
Some notations are a bit puzzling. For example, the open bracket below seems to need a term before the comma, but none is there. What is meant by this notation?

Alfven_speed(B, density[, ion, z_mean])

**Comment by namurphy** [comment]
I also noticed that there was a thing on [this page](https://plasmapy--1105.org.readthedocs.build/en/1105/formulary/parameters.html):

```
gyroradius(B, particle, *[, Vperp, T_i])
```
I'm not sure if there should be a comma or a space after the `*`.  In any case, we do think that this is a great improvement!

**Comment by rocco8773** [comment]
@rebeccaringuette and @namurphy When a the signature of a documented function shows parameters in square brackets, then those parameters are "optional" parameters.  This is the behavior of Sphinx's `autodoc` directives and `autosummary`.  So, in the two examples you two mentioned...

`Alfven_speed(B, density[, ion, z_mean])` --> optional parameters are `ion` and `z_mean`
`gyroradius(B, particle, *[, Vperp, T_i])` --> optional parameters are `Vperp` and `T_i`

You can see this throughout documentation for various Python packages.  For example, python's [`dir()`](https://docs.python.org/3/library/functions.html#dir) function has an optional single argument `object`.

**Review by StanczakDominik — APPROVED** [review]
Well, I've read the docs, fixed a few small typos. LveryGTM, I'd be in favor of a swift merge :)

**Review by namurphy — APPROVED** [review]

**Pull request merged**
